site/src/components/PaginationWidget/PaginationContainer.tsx

```diff
@@ -0,0 +1,214 @@
+            currentPage={paginationResult.currentPage}
+            pageSize={paginationResult.limit}
+            onPageChange={paginationResult.onPageChange}
+            hasPreviousPage={paginationResult.hasPreviousPage}
+            hasNextPage={paginationResult.hasNextPage}
+          />
+        )}
+      </div>
+    </div>
+  );
+};
+
+// Events to listen to for canceling queued scrolls
+const userInteractionEvents: (keyof WindowEventMap)[] = [
+  "click",
+  "scroll",
+  "pointerenter",
+  "touchstart",
+  "keydown",
+];
+
+/**
+ * Splitting this into a custom hook because there's a lot of convoluted logic
+ * here (the use case doesn't line up super well with useEffect, even though
+ * it's the only tool that solves the problem). Please do not export this; it
+ * should be treated as an internal implementation detail
+ *
+ * Scrolls the user to the top of the pagination container when the current
+ * page changes (accounting for old data being shown during loading transitions)
+ *
+ * See Pagination test file for all cases this is meant to handle
+ */
+function useScrollOnPageChange(
+  currentPage: number,
+  showingPreviousData: boolean,
+  autoScroll: boolean,
+) {
+  const scrollContainerRef = useRef<HTMLDivElement>(null);
+  const isScrollingQueuedRef = useRef(false);
+
+  // Sets up event handlers for canceling queued scrolls in response to
+  // literally any user interaction
+  useEffect(() => {
+    if (!autoScroll) {
+      return;
+    }
+
+    const cancelScroll = () => {
+      isScrollingQueuedRef.current = false;
+    };
+
+    for (const event of userInteractionEvents) {
+      window.addEventListener(event, cancelScroll);
+    }
+
+    return () => {
+      for (const event of userInteractionEvents) {
+        window.removeEventListener(event, cancelScroll);
+      }
+    };
+  }, [autoScroll]);
+
+  const scrollToTop = useEffectEvent(() => {
+    const newVerticalPosition =
+      (scrollContainerRef.current?.getBoundingClientRect().top ?? 0) +
+      window.scrollY;
+
+    // Not using element.scrollIntoView for testing reasons; much easier to mock
+    // the global window object
+    window.scrollTo({ top: newVerticalPosition, behavior: "instant" });
+    isScrollingQueuedRef.current = false;
+  });
+
+  // Reminder: effects always run on mount, no matter what's in the dependency
+  // array. Not doing anything on initial render because unconditionally
+  // scrolling and hijacking the user's page will feel absolutely awful
+  const isOnFirstRenderRef = useRef(true);
+  const syncPageChange = useEffectEvent(() => {
+    if (isOnFirstRenderRef.current) {
+      isOnFirstRenderRef.current = false;
+      return;
+    }
+
+    if (showingPreviousData) {
+      isScrollingQueuedRef.current = true;
+    } else {
+      scrollToTop();
+    }
+  });
+
+  // Would've liked to consolidate these effects into a single useLayoutEffect
+  // call, but they kept messing each other up when grouped together
```

Comment:
Do we know the reason?

Reply:
It's not because it's an unknown reason – it's more that there are two "sync cues" that the UI needs to account for, and trying to squeeze both of them into one layout effect just got a little unruly. It felt more manageable to set up one effect for the page changes, and one for fetching state, especially since they can change independently

I can rewrite the comment to make that more clear, though